Defmt feature not working

At the moment, `defmt::Format` is not implemented for error types which causes errors when compiling with the `defmt` feature.

This is probably easy to fix by adding some `#[cfg_attr(feature = "defmt", derive(defmt::Format))]`

I think it would be good to add CI tests that check every feature (either by enabling all features if possible or running `cargo check` for all available features one by one).

```
$ cargo check --no-default-features --features defmt
    Checking cfg-noodle v0.1.0 (/path/cfg-noodle)
error[E0277]: the trait bound `LoadStoreError<<F as ErrorType>::Error>: Format` is not satisfied
   --> src/intrusive.rs:634:13
    |
634 | /             error!(
635 | |                 "First writing attempt failed! Error: {}. Trying again...",
636 | |                 e
637 | |             );
    | |_____________^ the trait `Format` is not implemented for `LoadStoreError<<F as ErrorType>::Error>`
    |
    = help: the following other types implement trait `Format`:
              &T
              &mut T
              ()
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
            and 121 others
note: required by a bound in `defmt::export::fmt`
   --> /path/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/defmt-1.0.1/src/export/mod.rs:137:15
    |
137 | pub fn fmt<T: Format + ?Sized>(f: &T) {
    |               ^^^^^^ required by this bound in `fmt`
    = note: this error originates in the macro `error` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `cfg-noodle` (lib) due to 1 previous error
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -1,4 +1,4 @@
-name: Build and Test 
+name: Build and Test
 
 on:
   push:
@@ -15,8 +15,8 @@ jobs:
       - uses: actions/checkout@v4
       - uses: taiki-e/install-action@cargo-hack
       - name: Check each feature
-        run: cargo hack check --each-feature --no-dev-dep 
+        run: cargo hack check --each-feature --no-dev-deps
       - name: Build each feature
-        run: cargo hack build --each-feature --no-dev-dep 
+        run: cargo hack build --each-feature --no-dev-deps
       - name: Test each feature
-        run: cargo hack test --each-feature --no-dev-dep 
+        run: cargo hack test --each-feature --no-dev-deps

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -1,0 +1,22 @@
+name: Build and Test 
+
+on:
+  push:
+    branches: ["main"]
+  pull_request:
+    branches: ["main"]
+  workflow_dispatch:
+
+jobs:
+  miri:
+    name: "Check"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: taiki-e/install-action@cargo-hack
+      - name: Check each feature
+        run: cargo hack check --each-feature --no-dev-dep 
+      - name: Build each feature
+        run: cargo hack build --each-feature --no-dev-dep 
+      - name: Test each feature
+        run: cargo hack test --each-feature --no-dev-dep 

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -15,8 +15,8 @@ jobs:
       - uses: actions/checkout@v4
       - uses: taiki-e/install-action@cargo-hack
       - name: Check each feature
-        run: cargo hack check --each-feature --no-dev-deps
+        run: cargo hack check --each-feature --no-dev-deps --exclude-all-features --exclude-features arbitrary
       - name: Build each feature
-        run: cargo hack build --each-feature --no-dev-deps
+        run: cargo hack build --each-feature --no-dev-deps --exclude-all-features --exclude-features arbitrary
       - name: Test each feature
-        run: cargo hack test --each-feature --no-dev-deps
+        run: cargo hack test --features std

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -26,7 +26,7 @@ std = [
     "dep:simple_logger",
     "dep:tokio",
 ]
-defmt = ["dep:defmt"]
+defmt = ["dep:defmt", "sequential-storage/defmt-03"]
 arbitrary = ["dep:arbitrary"]
 
 [dependencies]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -41,7 +41,6 @@ maitake-sync = "0.2.1"
 mutex = { version = "1.0.0", features = ["impl-critical-section"] }
 
 pin-project = "1.1.10"
-thiserror = { version = "2.0.12", default-features = false }
 
 log = { version = "0.4.27", optional = true }
 simple_logger = { version = "5.0.0", optional = true }

--- a/src/error.rs
+++ b/src/error.rs
@@ -1,56 +1,44 @@
 //! Error types used for the list
-use embedded_storage_async::nor_flash::NorFlashError;
 use sequential_storage::Error as SecStorError;
 
 use crate::intrusive::{KEY_LEN, State};
 
 /// General error that is not specific to the flash implementation
-#[derive(thiserror::Error, Debug)]
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum Error {
     /// Deserializing node from the buffer failed
-    #[error("Deserialization failed!")]
     Deserialization,
     /// Serializing a node into the buffer failed
-    #[error("Serialization failed! Buffer too small.")]
     Serialization,
     /// Old version of the node was found. Returned when the node does not have the latest
     /// counter value.
-    #[error("Only an old instance of this StorageListNode was found")]
     OldData,
     /// The key hash already exists in the list
-    #[error("Duplicate key(-hash)!")]
     DuplicateKey,
     /// Node is in a state that is invalid at the particular point of operation.
     /// Contains a tuple of the node key hash and the state that was deemed invalid.
-    #[error("Invalid Node State! Key {:?}, {:?}", .0, .1)]
     InvalidState([u8; KEY_LEN], State),
 }
-
 
 // TODO @James: I have created this error because it uses the generic <F>. Adding generics
 // to the generall error struct was super awkward. Is there any better way of wrapping errors
 // from other crates when they require generics?
 
-#[derive(thiserror::Error, Debug)]
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 /// Errors during loading from and storing to flash.
-/// 
+///
 /// Sometimes specific to the flash implementation.
-pub enum LoadStoreError<F: NorFlashError> {
+pub enum LoadStoreError<F> {
     /// Writing to flash has failed. Contains the error returned by sequential storage.
-    #[error("Writing to flash has failed: {0:?}")]
     FlashWrite(SecStorError<F>),
     /// Reading from flash has failed. Contains the error returned by sequential storage.
-    #[error("Reading from flash has failed: {0:?}")]
     FlashRead(SecStorError<F>),
     /// Value read back from the flash during verification did not match serialized list node.
-    #[error("Value written to flash does not match serialized list node.")]
     WriteVerificationFailed,
     /// Recoverable error to tell the caller that the list needs reading first.
-    #[error(
-        "List needs to be read first because some node is in Initial state and cannot be written."
-    )]
     NeedsRead,
     /// Application-level error that occurred during list operations.
-    #[error(transparent)]
-    AppError(#[from] Error),
+    AppError(Error),
 }

--- a/src/intrusive.rs
+++ b/src/intrusive.rs
@@ -150,10 +150,10 @@ pub struct Node<T> {
 
     // We generally want this to be in the tail position, because the size of T varies
     // and the pointer to the `NodeHeader` must not change as described above.
-    // 
+    //
     // TODO @James: This is very difficult to access from tests but maybe we would
     // want to have an (test-)interface that lets us create a node with a specific payload
-    // such that we can directly de-/serialize nodes and compare the results in a 
+    // such that we can directly de-/serialize nodes and compare the results in a
     // unit test. Do you think that makes sense? Or is there a better way to
     // have unit tests for this?
     t: MaybeUninit<T>,
@@ -211,6 +211,7 @@ pub struct NodeHeader {
 /// ```
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum State {
     /// The node has been created, but never "hydrated" with data from the
     /// flash. In this state, we are waiting to be notified whether we can
@@ -630,11 +631,8 @@ impl<R: ScopedRawMutex> StorageList<R> {
         debug!("Attempt write_to_flash");
         // ... and try writing the list to flash one time.
         // If this fails, try again. If it fails again, return the error.
-        if let Err(e) = write_to_flash(&mut ls, serde_buf, flash).await {
-            error!(
-                "First writing attempt failed! Error: {}. Trying again...",
-                e
-            );
+        if write_to_flash(&mut ls, serde_buf, flash).await.is_err() {
+            error!("First writing attempt failed!");
 
             // Increase the counter for this list by two
             max_counter += Wrapping(1);
@@ -1559,7 +1557,7 @@ mod test {
                     embassy_futures::select::Either::First(_) => {
                         info!("worker task got needs_read signal");
                         if let Err(e) = list.process_reads(&mut flash, &mut read_buf).await {
-                            error!("Error in process_writes: {}", e);
+                            error!("Error in process_reads: {:?}", e);
                         }
                     }
                     embassy_futures::select::Either::Second(_) => {
@@ -1568,7 +1566,7 @@ mod test {
                             .process_writes(&mut flash, &mut read_buf, &mut serde_buf)
                             .await
                         {
-                            error!("Error in process_writes: {}", e);
+                            error!("Error in process_writes: {:?}", e);
                         }
 
                         info!("Wrote to flash: {}", flash.flash().print_items().await);

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -3,8 +3,8 @@
 #![warn(missing_docs)]
 #![deny(clippy::unwrap_used)]
 pub mod error;
-pub mod intrusive;
 pub mod flash;
+pub mod intrusive;
 
 #[allow(unused)]
 pub(crate) mod logging {
@@ -13,6 +13,9 @@ pub(crate) mod logging {
 
     #[cfg(feature = "defmt")]
     pub use defmt::*;
+
+    #[cfg(all(feature = "std", feature = "defmt"))]
+    compile_error!("Cannot enable both 'std' and 'defmt' features simultaneously");
 
     /// No-op macros when no logging feature is enabled
     #[cfg(not(any(feature = "std", feature = "defmt")))]

--- a/src/main.rs
+++ b/src/main.rs
@@ -42,7 +42,7 @@ async fn main() {
             .process_writes(&mut flash2, read_buf, serde_buf)
             .await
         {
-            error!("Error in process_writes: {}", e);
+            error!("Error in process_writes: {:?}", e);
         }
         info!("NEW WRITES: {}", flash2.flash().print_items().await);
     }

--- a/tests/integration_tests.rs
+++ b/tests/integration_tests.rs
@@ -116,7 +116,7 @@ fn worker_task<R: ScopedRawMutex + Sync>(
                 embassy_futures::select::Either3::First(_) => {
                     info!("worker task got needs_read signal");
                     if let Err(e) = list.process_reads(&mut flash, &mut read_buf).await {
-                        error!("Error in process_writes: {}", e);
+                        error!("Error in process_reads: {:?}", e);
                     }
                 }
                 // Handle write request from storage list
@@ -127,7 +127,7 @@ fn worker_task<R: ScopedRawMutex + Sync>(
                         .process_writes(&mut flash, &mut read_buf, &mut serde_buf)
                         .await
                     {
-                        error!("Error in process_writes: {}", e);
+                        error!("Error in process_writes: {:?}", e);
                     }
                 }
                 // Handle termination signal or timeout - exit the loop and return flash

